rust/noosphere-core/src/api/headers/ucan.rs

```diff
@@ -0,0 +1,90 @@
+use crate::data::Jwt;
+use anyhow::anyhow;
+use cid::Cid;
+use headers::{self, Header, HeaderName, HeaderValue};
+use once_cell::sync::Lazy;
+use ucan::{chain::ProofChain, crypto::did::DidParser, store::UcanJwtStore};
+
+static UCAN_NAME: Lazy<HeaderName> = Lazy::new(|| HeaderName::from_static("ucan"));
+
+/// A typed header for the `ucan` header, a tuple of [Cid] and [Jwt].
+/// Can represent a request with multiple `ucan` headers.
+///
+/// The values are **not** validated during parsing beyond legitimate
+/// looking [Cid]s and [Jwt]s here. Use [Ucan::as_proof_chain] to
+/// validate and construct a [ProofChain].
+pub struct Ucan(Vec<(Cid, Jwt)>);
+
+impl Header for Ucan {
+    fn name() -> &'static HeaderName {
+        &UCAN_NAME
+    }
+
+    fn decode<'i, I>(values: &mut I) -> Result<Self, headers::Error>
+    where
+        I: Iterator<Item = &'i HeaderValue>,
+    {
+        let mut ucans = vec![];
+        for header in values.by_ref() {
+            let value = header.to_str().map_err(|_| headers::Error::invalid())?;
+            let mut parts: Vec<&str> = value.split_ascii_whitespace().take(2).collect();
+
+            let jwt: Jwt = parts
+                .pop()
+                .ok_or_else(headers::Error::invalid)?
+                .to_string()
+                .into();
+            let cid: Cid = parts
+                .pop()
+                .ok_or_else(headers::Error::invalid)?
+                .try_into()
+                .map_err(|_| headers::Error::invalid())?;
+
+            ucans.push((cid, jwt));
+        }
+
+        Ok(Ucan(ucans))
+    }
+
+    fn encode<E>(&self, values: &mut E)
+    where
+        E: Extend<HeaderValue>,
+    {
+        for (cid, jwt) in self.0.iter() {
+            let value = HeaderValue::from_str(&format!("{} {}", cid, jwt)).unwrap();
+            values.extend(std::iter::once(value));
+        }
+    }
+}
+
+impl Ucan {
+    /// Construct a [ProofChain] from the collected `ucan` header
+    /// values. This is where validation of the data occurs.
```

Comment:
No logic changes here if this `Header` implementation correctly parses *all* ucan headers before handling here (hence the `Vec<(Cid, Jwt)>`), which all get written to the store before proof chain is constructed, IIUC

Reply:
Added spec link and a note that we are not currently adhering to the 0.3.0 spec with a single `ucans` header, rather than multiple `ucan` headers. #708